lib/spack/spack/solver/asp.py

```diff
@@ -3393,7 +3406,8 @@ def _develop_specs_from_env(spec, env):
         assert spec.variants["dev_path"].value == path, error_msg
     else:
         spec.variants.setdefault("dev_path", spack.variant.SingleValuedVariant("dev_path", path))
-    spec.constrain(dev_info["spec"])
+
+    assert spec.satisfies(dev_info["spec"])
```

Comment:
This 3 line diff addresses https://github.com/spack/spack/pull/41049#issuecomment-2096014945:

This uncovers a discrepancy with `constrain`, where it's unwilling to add flags (for `x.constrain(y)`, the implementation of `constrain` explicitly forbids that `x` contains flags not in `y`). I decided that WRT this PR the solution is that we shouldn't be applying more constraints after the solver completes, and skipped applying that function.

In the process of reproducing that issue (`test_dev_mix_flags`) I also found that propagation would generally not mix with other forms of flags because of a rule in the concretizer, I'll add a separate comment for that.

Reply:
Not sure I entirely follow the comment above :thinking:

Reply:
* Flags should not be added at this stage (the concretizer should have combined them already)
* Therefore `constrain` is not the right method to use here
* By removing it, I sidestepped an issue where`.constrain` is not conceptually synced with the changes to `.satisfies` in this PR: for `x.constrain(y)`, the current definition of `FlagMap.constrain` (which is unaltered in this PR) will fail if x contains a flag that is not in y
* I figured this should be addressed, but I didn't want to handle that in this PR

Is that more clear?

Reply:
I agree with the first couple of points, less with leaving `constrain` in an inconsistent state. I can help with that, in case.

Reply:
Should be resolved as of 3d75fbbe43b772566aeb9b2ec8ccd366c1ccacef/https://github.com/spack/spack/pull/41049#discussion_r1724045046